### DISCUSSION TITLE
Bucket4j JDBC integration

### DISCUSSION THREAD

**Discussion by BarracudaX** [vladimir-bukhtoyarov/bucket4j#439] [Q&A]
Hello! I am new to bucket4j. I want to use it in my spring boot application that uses Spring MVC to use rate limiting on my APIs.

I have the following problem: I need to rate limit some APIs based on user role/subscription(for example, free,premium etc) but also some APIs need rate limiting based on IP. Is it possible to do this with bucket4j?

After reading the documentation, I assume I can use bucket id(long) to associate it with users - assign each user bucket id. Correct me if I am wrong. This should work for API rate limiting based on user's subscription/role. But what about IP address? How can I store it in the table if the primary key for bucket is id(long)? Should I find a way to convert IP to Long or can I just add additional nullable fields to the buckets table for ip address? Is there a better solution?

**Comment by vladimir-bukhtoyarov** [comment]
You are free to use any type as primary key, for example:
```java
SQLProxyConfiguration<String> sqlConfiguration = SQLProxyConfiguration.builder()
            .withPrimaryKeyMapper(PrimaryKeyMapper.STRING)
            .withTableSettings(tableSettings)
            .build(dataSource);
ProxyManager<String> proxyManager = new PostgreSQLadvisoryLockBasedProxyManager<>(sqlConfiguration)
```

**Reply by BarracudaX** [reply]
Okay, thanks. I will then use two different configurations and two different proxy managers.

I have another question about limits. In documentation it says that a bucket can have more than one limit, but I am not sure if I understand how it works then. If I have a bucket with two limits: one has 100 tokens with refill happening every 10 seconds and another limit has 10 tokens with refill happening every 1 second, then when I consume from a bucket 1 token, how will it behave? Will it try to consume from first limit and only go to the second if first limit is exhausted?

Update: I think I found answer in javadocs of Bandwidth class " bucket will handle all bandwidth in strongly atomic way. Strongly atomic means that token will be consumed from all bandwidth or from nothing, in other words any token can not be partially consumed."
